src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java

```diff
@@ -0,0 +1,107 @@
+package core.basesyntax.service;
+
+import static core.basesyntax.service.RegistrationServiceImpl.USER_MIN_ACCESSIBLE_AGE;
+import static core.basesyntax.service.RegistrationServiceImpl.WHITE_SPACE;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import core.basesyntax.exceptions.UserRegistrationException;
+import core.basesyntax.model.User;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class RegistrationServiceImplTest {
+    private static RegistrationServiceImpl registrationService;
+    private static final int USER_AGE_LESS_18_NOT_OK = 17;
+    private static final int USER_AGE_MORE_THAN_18_OK = 30;
+    private static final String USER_LOGIN_LENGTH_LESS_6_SYMBOLS_NOT_OK = "Bob";
+    private static final String USER_LOGIN_LENGTH_6_SYMBOLS_OK = "Bob123";
+    private static final String USER_PASSWORD_LENGTH_LESS_6_SYMBOLS_NOT_OK = "bob";
+    private static final String USER_PASSWORD_LENGTH_6_SYMBOLS_OK = "bob123";
+
+    @BeforeAll
+    static void beforeAll() {
+        registrationService = new RegistrationServiceImpl();
+    }
+
+    @ParameterizedTest
+    @MethodSource("register_userData_Ok")
+    void register_addUserToDatabase_Ok(User input) {
+        User actualResult = registrationService.register(input);
+        assertEquals(input.getLogin(), actualResult.getLogin());
+        assertEquals(input.getPassword(), actualResult.getPassword());
+        assertEquals(input.getAge(), actualResult.getAge());
+    }
+
+    @ParameterizedTest
+    @MethodSource("register_userData_NotOk")
+    void register_addUserToDatabase_notOk(User input) {
+        assertThrows(UserRegistrationException.class, () -> registrationService.register(input));
+    }
+
+    private static Stream<Arguments> register_userData_Ok() {
+        return Stream.of(
+                Arguments.of(new User(
+                        USER_LOGIN_LENGTH_6_SYMBOLS_OK,
+                        USER_PASSWORD_LENGTH_6_SYMBOLS_OK,
+                        USER_MIN_ACCESSIBLE_AGE)),
+                Arguments.of(new User(
+                        USER_LOGIN_LENGTH_6_SYMBOLS_OK,
+                        USER_PASSWORD_LENGTH_6_SYMBOLS_OK,
+                        USER_AGE_MORE_THAN_18_OK))
+        );
```

Comment:
I'm far from streams, but how are adding 2 users with the same login? We expect an exception for this case.

Reply:
This construct provides us with a stream of new users (eg a list of different users). Each part of the stream that starts with "Arguments.of" is a separate user. And we add them one by one to test the method and test adding a new user with the specified data to the repository. We don't add them all at once. It's like we can create every test method for every test case. But this design allows you to accumulate all test cases it in one place.